# Project Introduction
This project explores the data from UCI Machien Learning posted up on Kaggle. 

In [1]:
# Imports 
%matplotlib inline

# General Python libraries
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd
import re
import time

# Feature Extraction & Unsupervised Libraries
from sklearn.feature_extraction.text import *
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.mixture import GMM

from matplotlib.colors import LogNorm

# Learning/Model Libraries
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import BernoulliNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import GridSearchCV
# Add trees libraries

# Evaluation Libarires
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import classification_report

## Data Processing

In [2]:
# Read in the raw data, process and save into different folder
RAWDATA_PATH = '../data/raw/UCIHumanActivity/'
df_rtrain = pd.read_csv(f'{RAWDATA_PATH}train.csv')
df_rtrain.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.983111         -0.913526         -0.995112         -0.983185   
1         -0.975300         -0.960322         -0.998807         -0.974914   
2         -0.967187         -0.978944         -0.996520         -0.963668   
3         -0.983403         -0.990675         -0.997099         -0.982750   
4         -0.980817         -0.990482         -0.998321         -0.979672   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X    ...     \
0         -0.923527         -0.934724    ...      
1         -0.957686         -0.943068    ...      
2         -0.977469         -0.938692    ...      
3         -0.989302         -0.938692    ...      
4         -0.990441         -0.942469    ...      

   fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
0                        -0.710304                    -0.112754   
1                        -0.861499                     0.053477   
2                        -0.760104                    -0.118559   
3                        -0.482845                    -0.036788   
4                        -0.699205                     0.123320   

   angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
0                              0.030400                         -0.464761   
1                             -0.007435                         -0.732626   
2                              0.177899                          0.100699   
3                             -0.012892                          0.640011   
4                              0.122542                          0.693578   

   angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
0                             -0.018446             -0.841247   
1                              0.703511             -0.844788   
2                              0.808529             -0.848933   
3                             -0.485366             -0.848649   
4                             -0.615971             -0.847865   

   angle(Y,gravityMean)  angle(Z,gravityMean)  subject  Activity  
0              0.179941             -0.058627        1  STANDING  
1              0.180289             -0.054317        1  STANDING  
2              0.180637             -0.049118        1  STANDING  
3              0.181935             -0.047663        1  STANDING  
4              0.185151             -0.043892        1  STANDING  

[5 rows x 563 columns]

In [7]:
print(df_rtrain.shape)
print(df_rtrain.dtypes[:5])

(7352, 563)
tBodyAcc-mean()-X    float64
tBodyAcc-mean()-Y    float64
tBodyAcc-mean()-Z    float64
tBodyAcc-std()-X     float64
tBodyAcc-std()-Y     float64
dtype: object


## Observations on Training Data
* There are 7352 examples in the training set
* There are 563 columns 
* Activity, the last column, is the label (Y) 
* The subject is identified by a number
* All other columns are float64 datatypes

## Next Steps
* Some histograms
    * On activities
    * On the subject
* Separate out X and Y (features and labels) as well as the participants
* Consider filtering now to the original, raw features since 563 is A LOT of features 
* Consider creating a validation set

In [12]:
print('Activity Counts: ')
df_rtrain['Activity'].value_counts()

Activity Counts: 


LAYING                1407
STANDING              1374
SITTING               1286
WALKING               1226
WALKING_UPSTAIRS      1073
WALKING_DOWNSTAIRS     986
Name: Activity, dtype: int64

In [13]:
print('Participant Counts: ')
df_rtrain['subject'].value_counts()
# 22 participants included with 281 to 409 measurements

Participant Counts: 


25    409
21    408
26    392
30    383
28    382
27    376
23    372
17    368
16    366
19    360
1     347
29    344
3     341
15    328
6     325
14    323
22    321
11    316
7     308
5     302
8     281
Name: subject, dtype: int64

In [39]:
df_subj_act = df_rtrain[['subject', 'Activity']]
df_pivot = pd.pivot_table(df_subj_act, index=['subject'], columns=['Activity'], aggfunc = 'count')
df_pivot


TypeError: pivot_simple() got multiple values for argument 'index'

In [8]:
df_rtest = pd.read_csv(f'{RAWDATA_PATH}test.csv')
df_rtest.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.257178          -0.023285          -0.014654         -0.938404   
1           0.286027          -0.013163          -0.119083         -0.975415   
2           0.275485          -0.026050          -0.118152         -0.993819   
3           0.270298          -0.032614          -0.117520         -0.994743   
4           0.274833          -0.027848          -0.129527         -0.993852   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.920091         -0.667683         -0.952501         -0.925249   
1         -0.967458         -0.944958         -0.986799         -0.968401   
2         -0.969926         -0.962748         -0.994403         -0.970735   
3         -0.973268         -0.967091         -0.995274         -0.974471   
4         -0.967445         -0.978295         -0.994111         -0.965953   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X    ...     \
0         -0.674302         -0.894088    ...      
1         -0.945823         -0.894088    ...      
2         -0.963483         -0.939260    ...      
3         -0.968897         -0.938610    ...      
4         -0.977346         -0.938610    ...      

   fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
0                        -0.705974                     0.006462   
1                        -0.594944                    -0.083495   
2                        -0.640736                    -0.034956   
3                        -0.736124                    -0.017067   
4                        -0.846595                    -0.002223   

   angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
0                              0.162920                         -0.825886   
1                              0.017500                         -0.434375   
2                              0.202302                          0.064103   
3                              0.154438                          0.340134   
4                             -0.040046                          0.736715   

   angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
0                              0.271151             -0.720009   
1                              0.920593             -0.698091   
2                              0.145068             -0.702771   
3                              0.296407             -0.698954   
4                             -0.118545             -0.692245   

   angle(Y,gravityMean)  angle(Z,gravityMean)  subject  Activity  
0              0.276801             -0.057978        2  STANDING  
1              0.281343             -0.083898        2  STANDING  
2              0.280083             -0.079346        2  STANDING  
3              0.284114             -0.077108        2  STANDING  
4              0.290722             -0.073857        2  STANDING  

[5 rows x 563 columns]

In [9]:
print(df_rtest.shape)
print(df_rtest.dtypes[:5])

(2947, 563)
tBodyAcc-mean()-X    float64
tBodyAcc-mean()-Y    float64
tBodyAcc-mean()-Z    float64
tBodyAcc-std()-X     float64
tBodyAcc-std()-Y     float64
dtype: object
